### ISSUE TITLE
Request: Add Golang API bindings for C-API TF_AddGradients()

### ISSUE THREAD

**Issue by NMRobert** [tensorflow/tensorflow#21404]
Hi,

currently the Golang bindings to Tensorflow (via the C API) allow me to define a graph, do most computations, load a trained model, and perform inference. However, it would be very helpful to be able to obtain gradients aswell. It looks like the C-API now has the TF_AddGradients call which makes it possible to get grads for some (but not all) operations. 

Adding a binding to this call in the Go API would allow me to use the C API through golang as a general autodiff library, which would help with some projects (such as the implementation of gradient-based stochastic samplers in Go like SGHMC). If this is actually easier than it looks, any pointers on how I might attempt it would be welcomed. Thanks!

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce
Mobile device

**Comment by shivaniag** [comment]
@alextp Do you have any pointers in this?

**Comment by alextp** [comment]
I'm tempted to mark this as contributions welcome; it should be straightforward with the way the go bindings work. Asim, does that make sense?

**Comment by asimshankar** [comment]
Yup, it should be relatively straightforward and can be done as it was similarly done for Java recently (commit fac56f9c9ab58fe7406a826683559de4cef85637).

Marking as "Contributions Welcome" in case someone wants to get to it before I do :)

**Comment by sanketloke** [comment]
@asimshankar I would like to try this if no else is. Would send a PR by next week.

**Comment by asimshankar** [comment]
@sanketloke : Sure thing, looking forward to it.

**Comment by Cibifang** [comment]
After a week and still not fixed, it seems that  no one continue this issue, so I just submit a pull request.

Wish it's not very abrupt

**Comment by marpaia** [comment]
Is there still work to do here or did #21895 wrap this up?

**Comment by asimshankar** [comment]
Thanks for linking this up @marpaia . Yeah #21895 should have closed this (and if there are missing features, a new issue can/should be filed).

@ymodak : Could you close this? Thanks!
